Optim wip - General fixes, SharedImage, & Weight Visualization

ProGamerGov opened the pull request:
Just a quick update that fixes a few bugs, adds SharedImage (`lowres_tensor` from Lucid / Lucent), two new transforms for normalization, and weight visualization tools.

The two new transforms are extremely simple and could be done in other ways like using torchvision's lambda transform, but having them readily accessible should make it easier for Captum's users. In the future we may be able to expand the functionality of these two transforms. The two new transforms are based on discussions from the project doc regarding how to support models with different normalization requirements.

* Added SharedImage parameterization that's based off of Lucid / Lucent's lowres_tensor. I've included a citation for the paper / article that it comes from.  I've tested SharedImage pretty extensively and it works correctly.

* Added ScaleInputRange transform for use with models that have an input range other than 0,1. 

* Added RGBToBGR transform for converting RGB inputs to BGR. 

* Added tests for SharedImage, ScaleInputRange, RGBToBGR.

* Some fixes to alpha channel support and additional asserts.

* Fix NumPy FFTImage so that it outputs almost the exact the same values as the PyTorch version.

* Added preliminary tools for circuit research. The code is based off of a research article that is set to be released in the new few days.

* Added `n_channels_to_rgb` function that's based `collapse_channels` from Lucid / Lucent.

* Add tool for weight channel dimensionality reduction.

* Added tutorial for weight visualization. More complex tutorial steps will be implemented in a future PR after https://github.com/pytorch/captum/pull/552 is merged, because https://github.com/pytorch/captum/pull/552 is waiting on this PR.

* Added tutorial for creating custom transforms, loss functions, and input parameterizations.

* Fixed some mistakes in the InceptionV1 model class.

* The use of the tqdm library is now optional.

The usage of SharedImage does seem a bit awkward right now, but it's written in such a way that we can easily change it in the future.

```
image = optimviz.images.NaturalImage(...)
shapes = # a tuple or tuple of tuples of tensor shapes with a length of 2-4 (NCHW), (CHW), (HW)
image.parameterization = optimviz.images.SharedImage(shapes=shapes, parameterization=image.parameterization)
```

ProGamerGov commented:
@NarineK This is a relatively small update, and it's ready for review as I'm not adding or changing anything else right now.

ProGamerGov commented:
There may be something off with Ludwig's code:

```
import torch

test_tensor = torch.randn(1, 3, 512, 405)
fft_tensor = torch.rfft(test_tensor, signal_ndim=2)

# Ludwig's FFTImage scaled_spectrum size code
h, w = test_tensor.size(2), test_tensor.size(3)
wadd = 2 if w % 2 == 1 else 1
expected_size = torch.ones([3, h, w // 2 + wadd])

print(expected_size.size(2), fft_tensor.size(3)) # 204 203
out = fft_tensor / expected_size # Fails due to size mismatch
```

ProGamerGov commented:
@NarineK Changing `np.fft.rfftn` to `np.fft.irfftn` in the NumPy FFTImage was causing a `Received "killed" signal` message in the CircleCI tests. Example: https://app.circleci.com/pipelines/github/pytorch/captum/2486/workflows/ef6cfb7c-3ce1-4ba6-b1b4-ac29a1356398/jobs/11988

I'm not sure if you can see a more in-depth message or anything like that? But I have zero idea why that change breaks CircleCI. For the time being I've reverted the numpy FFTImage class back to how it was before so the tests pass.

ProGamerGov commented:
@NarineK I know I said that I wasn't adding anything else to this PR, but I finished setting up / adding the channel reducer and expanded weights code sooner than expected! I don't plan to add anything else now.

NarineK commented on captum/optim/_core/objectives.py:
```diff
@@ -147,23 +150,31 @@ def optimize(
     """StopCriteria generator that uses number of steps as a stop criteria.
     Args:
         n (int):  Number of steps to run optimization.
+        show_progress (bool):  Whether or not to show progress bar.
```
nit: show_progress (bool, optional)
      Default: True

ProGamerGov replied:
Okay, I've made the change!

NarineK commented:
Thank you for working on this PR @ProGamerGov! Since it is a bigger PR, I am reviewing it step by step. Here are my comments related to `CustomModules_OptimViz`. I'll provide review comments for the rest of the code soon.

Thank you for creating Custom module's optimization tutorial. I think it is very helpful for users who want to use their own custom components.

 - Since there are multiple optim viz. related tutorials I think that it makes sense to move them all under one: optimviz folder.
   What do you think ?

- The custom transform that you describe is for image and I think that it would be good to mention in the comments that this is
   for image where we expect input to have `(B, C, W, H)` shape.

- nit: `... the base the loss class ...` -> `... the base loss class ...` ?

- nit: In the example of CustomLoss when you call, `target_activations[self.target]` you might want to make sure that target is a
  Module and not a list because the type hint can confuse people ? 

- CustomInput
1. For the `CustomInput` we might also want to emphasize in the description that this class is designed for image use cases
2. nit: We might want to move `assert size is not None` under `if init is None` because otherwise it is not used
3. nit type hint for size is missing

- Usually we also strip this kind of out. It would be good to remove it in order to make it look cleaner:
```
/usr/local/lib/python3.6/dist-packages/torch/tensor.py:775: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:848.)
  return super(Tensor, self).refine_names(names)
... 
```

NarineK commented:
@ProGamerGov, I see that Miscellaneous_OptimViz.ipynb got also recently added. Since this adds +1500 lines, I think that it would be good to move this to a separate PR so that we can move quicker.

ProGamerGov commented:
@NarineK Okay, I'll pull it out and save it for a future PR!

NarineK reviewed (commented):
> @NarineK Okay, I'll pull it out and save it for a future PR!

Thank you! Let's move it into a separate PR!

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +364,95 @@ def forward(self) -> torch.Tensor:
+
 class NaturalImage(ImageParameterization):
     r"""Outputs an optimizable input image.
 
```
The `set_image` functions come from Ludwig's original code, and they aren't being used anywhere currently. I can remove them if you want.

NarineK replied:
If it is not used we can remove it.

ProGamerGov commented on captum/optim/_core/objectives.py:
```diff
@@ -147,23 +150,31 @@ def optimize(
+        pbar = tqdm(total=n, unit="step")
 
     def continue_while(step, obj, history, optim) -> bool:
         if len(history) > 0:
```
Yeah, `False` is returned when we want to stop optimization. In the case of n_steps, `False` is returned when we reach the specified number of iterations:

```
        if step < n:
            if show_progress:
                pbar.update()
            return True
        else:
            if show_progress:
                pbar.close()
            return False
```

NarineK replied:
yes, I was thinking `if len(history) == 0` independent of `step`

ProGamerGov commented on captum/optim/_core/objectives.py:
```diff
@@ -147,23 +150,31 @@ def optimize(
 
 
-def n_steps(n: int) -> StopCriteria:
+def n_steps(n: int, show_progress: bool = True) -> StopCriteria:
```
The `n_steps` function and the `StopCriteria` type hint are both from Ludwig's original code. @greentfrapp and I will have to look into creating other stop criteria functions, and then it should become more clear if we need additional inputs for stopping criteria functions.

NarineK replied:
Sounds great! It would be good to refactor it and make it clearer. We can also do it in a separate PR.

ProGamerGov replied:
Yeah, I think it would be best to focus on improving stop criteria in a future PR.

ProGamerGov commented:
@NarineK I hope it is okay that I just added a rather small and simple function for dimensionality reduction. It should hopefully it make it easier to understand why I created a separate `reducer.py` file instead of just adding it to another file. It's also the second major way dimensionality reduction is handled for visualization tasks.

NarineK commented on captum/optim/_param/image/transform.py:
```diff
@@ -214,6 +215,30 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        assert x.dim() == 4
+        return x[:, [2, 1, 0]]
```
The dimension of channel is the 2nd dimension , right ? We expect bsz x channel x height x width ? 
Perhaps we can document that ?

ProGamerGov replied:
It's a transform for image tensors, so I expect that they'll be in the form of NCHW. I've added an assert to check that the number of channels is 3.

NarineK commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+        return A
+
+    def interpolate_tensor(
+        self, x: torch.Tensor, size: InitSize, batch: int, channels: int
```
nit: instead of passing size it might make more clear to pass hight and width.

ProGamerGov replied:
Yeah, that would probably make it easier to understand. I'll also order the input variables as BCHW as well.

NarineK commented on tests/optim/param/test_images.py:
```diff
@@ -242,5 +246,268 @@ def test_laplacianimage_random_forward(self) -> None:
+        )
+        test_tensor = image_param.forward()
+
+        self.assertEqual(test_tensor.dim(), 4)
```
Could we potentially also assert the shapes of `image_param.shared_init` and `image_param.offset` ?

ProGamerGov replied:
I've added asserts for `image_param.shared_init` in the latest commit, and I'll add the `image_param.offset` tomorrow.

NarineK reviewed (commented):
Posting comments for SharedImage, 2 new transforms and related test cases. Comments about circuits, related notebooks etc, I'll post in a separate review batch.

In terms of the awkward behavior  in SharedImage that you mentioned - I think that we can look deeper into offset and interpolation computations. Ideally we should be able to generate something described in: 
https://colab.research.google.com/github/tensorflow/lucid/blob/master/notebooks/differentiable-parameterizations/aligned_interpolation.ipynb#scrollTo=3HE5cCSlZrwa

I think that it would be great to add other new functionalities and new features in separate smaller PRs so that it is easy to review and also we can distinctly trace back the history and see when were things added in cases we want to debug some issues.

NarineK commented:
> @NarineK I hope that it's okay that I just added a rather small and simple function (shouldn't take much time to review I hope as it's basically only 30 lines) for dimensionality reduction.

Sounds good! Thank you :)

ProGamerGov commented on tests/optim/param/test_images.py:
```diff
@@ -242,5 +246,268 @@ def test_laplacianimage_random_forward(self) -> None:
         self.assertEqual(test_tensor.size(3), size[1])
 
 
+class TestSharedImage(BaseTest):
```
No, I don't believe there is any use case where shapes is None? I've added an assert to SharedImage.

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+    ) -> None:
+        super().__init__()
+        A = []
+        shared_shapes = [shapes] if type(shapes[0]) is not tuple else shapes
```
I've added an assert to check that shapes is not None.

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+        x = F.interpolate(x, size=size, mode=mode)
+        x = x.squeeze(0) if len(size) == 3 else x
+        if x.size(0) != batch:
+            x = x.permute(1, 0, 2, 3)
```
Normally F.interpolate doesn't resize the batch dimension, so I move the batch dimension to the channel dimension's location so that it gets resized. I'm not 100% sure this is the best way of resizing the batch dimension bilinearly, but it works for now. I'll add a simple comment about it.

ProGamerGov commented on tests/optim/param/test_images.py:
```diff
@@ -242,5 +246,268 @@ def test_laplacianimage_random_forward(self) -> None:
+        )
+
+        test_x_list = [torch.ones(*size) for x in range(size[0])]
+        output_A = image_param.apply_offset(test_x_list, size)
```
In the case of this test, there is only one set of shared_shapes specified. This means that there's only one shared tensor, and thus the apply_offset function will return a list containing the offset tensor. The size of output_A is dependent on how many sets of shared_shapes are defined.

NarineK replied:
I think in this case I was trying to refer to `apply_offset` method. When I run it in the notebook I see that:

```
self.offset =  [[2, 3, 4, 5]]
len(x_list) = 4 and x_list[i].shape = torch.Size([4, 3, 224, 224])
```
It looks like `self.offset` and `x_list` do not have the same length.

ProGamerGov replied:
Ah, I think I see the issue now. I've made some changes that should have the actual desired behavior that we are looking for!

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+                offset = [([0] * (4 - len(offset))) + list(offset)] * n
+        else:
+            offset = [[offset] * 4] * n
+        offset = [[int(o) for o in v] for v in offset]
```
Yeah, I guess we should check that they are all integers rather than trying to correct user mistakes behind the scenes.

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+        for shape in shared_shapes:
+            assert len(shape) >= 2 and len(shape) <= 4
+            shape = ([1] * (4 - len(shape))) + list(shape)
+            batch, channels, height, width = shape[0], shape[1], shape[2], shape[3]
```
Yes it does work, and I've made the change!

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+    def __init__(
+        self,
+        shapes: Union[Tuple[Tuple[int]], Tuple[int]] = None,
+        parameterization=None,
```
I've added the ImageParametrization type hint to all applicable locations!

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+                assert len(offset) == n and all(len(t) == 4 for t in offset)
+            else:
+                assert len(offset) >= 1 and len(offset) <= 4
+                offset = [([0] * (4 - len(offset))) + list(offset)] * n
```
The offset list is for each dimension of an NCHW tensor, and it looks like F.pad will pad the H and W dimensions, and possibly the C dimension as well?

NarineK replied:
Yeah, based on this examples, we need to make sure that the paddings are added to correct dimensions. We are intending to pad all  C, H and W if necessary, right ?

```
t4d = torch.empty(3, 3, 4, 2)
p3d = (0, 1, 2, 1, 3, 3) # pad by (0, 1), (2, 1), and (3, 3)
```

ProGamerGov replied:
The user can choose which dimensions (B, C, H, and / or W) to offset. So, I guess we need to pad all 4 dimensions? I may be misunderstanding what your saying as well, as I am a bit tired today. I'll have to dig into this deeper tomorrow as I'm trying to have healthier sleep schedule (had years of terrible insomnia up until recently).

NarineK replied:
Oh, it's not urgent. Take your time and get good sleep :) 
Yeah, I guess we can have offsets for all 4 dimensions.
https://github.com/tensorflow/lucid/blob/master/lucid/optvis/param/lowres.py#L63
Also, `tf.pad` can have a different signature and argument meaning than `torch.pad`. I'm not familiar with `tf.pad`.
https://github.com/tensorflow/lucid/blob/master/lucid/optvis/param/lowres.py#L63

ProGamerGov replied:
I think that I've figured out the correct padding now for all 4 dimensions!

ProGamerGov replied:
I had to use constant padding though, so I'll have to see about getting reflect padding working in a future PR.

ProGamerGov replied:
I've implemented reflection padding for all 4 dimensions, so now the padding should be correct! I had to add a new function though, and I wasn't 100% sure where I should put it (it's currently in model utils).

ProGamerGov commented:
@NarineK Yeah, I think we'll be able to make SharedImage less awkward in future PRs! I'll also try to separate new functionality groups into different PRs in the future now that we've got the base code working well.

ProGamerGov commented on captum/optim/_param/image/images.py:
```diff
@@ -364,6 +354,95 @@ def forward(self) -> torch.Tensor:
+    """
+    Share some image parameters across the batch.
+    Mordvintsev, et al., "Differentiable Image Parameterizations", Distill, 2018.
+    https://distill.pub/2018/differentiable-parameterizations/
```
Okay, I've added some more info to improve the description.

ProGamerGov commented:
CircleCI seems to have stopped running it's tests on each new commit and I'm not sure why.

ProGamerGov commented:
CircleCI tests don't seem to be showing up for my latest commits.

ProGamerGov commented:
CircleCI tests only seem to appear for my latest commit when I comment for some reason.

ProGamerGov commented:
CircleCI tests only seem to appear for my latest commit when I comment for some reason. I then have to comment again if I add a new commit.

ProGamerGov commented:
CircleCI tests only seem to appear for my latest commit when I comment for some reason. I then have to comment again if I add a new commit. If I refresh the page, they disappear again. They also disappear if a test fails or succeeds.

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

ProGamerGov commented:
t

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,72 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+
+def collect_activations(
+    model,
```
@ProGamerGov, Is this function and `ActivationCatcher` called anywhere except test cases ? Originally, as I checked out the code, it wasn't there

ProGamerGov replied:
It's used in the circuits file. The `ActivationCatcher` class exists to mimic how `InputOptimization` works so that it can reuse stuff, and so that notebook hook bug is prevented (hooks remain active or stop working for some reason). The `collect_activations` function makes it possible to use activation catch in one line of code.

ProGamerGov replied:
https://github.com/ProGamerGov/captum/blob/optim-wip/captum/optim/_utils/circuits.py

ProGamerGov replied:
The expanded weights function is use pretty heavily by the weight visualization notebook.

ProGamerGov replied:
@NarineK The expanded weights function is used pretty heavily by the weight visualization notebook, and the function itself required collecting activations in a way that the existing code didn't support easily.

NarineK replied:
I see - thank you! I haven't started reviewing circuits yet. I'll take a look into it. Originally as I checked out the code couple days ago, I could see `ActivationCatcher` but `collect_activations ` isn't there. I need to rebase...

NarineK commented on tests/optim/utils/test_model_utils.py:
```diff
@@ -9,6 +9,20 @@
+        x = torch.ones(1, 3, 4, 4)
+        padding = [2] * 8
+        x_out = model_utils.pad_reflective_a4d(x, padding)
+        self.assertEqual(list(x_out.shape), [4, 7, 8, 8])
```
why is the dimension [4, 7, 8, 8] here ? 
If we compare it with the constant equivalent then we get: torch.Size([5, 7, 8, 8]). 1 + 2 + 2 for the batch size dimension
```
torch.nn.functional.pad(x, padding, "constant", 0).shape
```

NarineK commented on tests/optim/utils/test_model_utils.py:
```diff
@@ -9,6 +9,20 @@
 
+class TestPadReflectiveA4D(BaseTest):
+    def test_pad_reflective_a4d(self) -> None:
+        x = torch.ones(1, 3, 4, 4)
```
Instead of using ones it might be better to use: x = torch.range(1, 3 * 4 * 4).view(1, 3, 4, 4) then you could also assert the values.

ProGamerGov replied:
`torch.range` is being depreciated, so this should be more backwards compatible: 

```
x = torch.arange(1,(3*4*4)+1).view(1, 3, 4, 4).float()
```

ProGamerGov replied:
We also might have to make batch and channel offsets raise a not implemented error for the time being, if I can't figure out how to fix the padding function.

ProGamerGov replied:
Okay, I've added the improved code! It works with everything but a channel or batch size of 2, and I compare the outputs with NumPy in the tests.

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,72 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = torch.cat([x, x.flip([2])[..., 0 : padding[2], :]], dim=2)
+    if padding[3] != 0:
+        x = torch.cat([x.flip([2])[..., -padding[3] :, :], x], dim=2)
+
```
for the hight and width we can probably use built-in function:
torch.nn.functional.pad(x, padding, "reflect") 

As I was trying it out I noticed that some elements get repeated:
```
... [22., 21., 21.,  ..., 24., 24., 23.] ...
```
I think that it would make sense to test the code for a tensor generated with torch.arange and  asserting actual values.

ProGamerGov replied:
Looks like repeating the boundary elements is desirable as that's what TensorFlow's symmetric padding does: https://github.com/pytorch/pytorch/issues/46240

NarineK replied:
@ProGamerGov, symmetric looks like replication padding but it is not implemented the way it is described in this article.
https://www.machinecurve.com/index.php/2020/02/07/what-is-padding-in-a-neural-network/

I think you can compare your implementation with the numpy variant to make sure that it works correctly
np.pad(a, 2, mode='symmetric')

ProGamerGov replied:
I think for the moment reflection padding should be okay. I modified the reflection padding function so that it works for all values, except an input size of 2 for channels and batch. https://github.com/ProGamerGov/captum/blob/optim-wip-patch-1/captum/optim/_utils/models.py#L160-L219

I'm currently trying to figure out how it's supposed to work with a size of 2.

ProGamerGov replied:
I compared my improved function with NumPy's reflection padding, and it works exactly the same (minus a batch or channel size of 2)!

NarineK replied:
I see - Are we supporting symmetric padding as in lucid or the reflection padding ?https://github.com/tensorflow/lucid/blob/master/lucid/optvis/param/lowres.py#L63

ProGamerGov replied:
Lucent uses reflection padding in place of symmetric padding, so for the moment I think we can do reflection padding. In a future PR, we can implement symmetric padding as an option for SharedImage.

ProGamerGov replied:
I've also reached out to Chris to double check that reflection padding is okay to use.

NarineK replied:
@ProGamerGov, let's keep the conversation in the PR thread in case other people look into this PR as well - I've received some of the messages as email response.

NarineK replied:
@ProGamerGov wrote: 

>>> I've also reached out to Chris to double check that reflection padding is okay to use.

Sounds good, thank you!

ProGamerGov replied:
@NarineK I haven't sent any emails, so I assume you mean the comments I deleted? I deleted some comments because I felt like they either didn't make sense, were completely wrong, or other stuff of that nature.

NarineK reviewed (commented):
Thank you @ProGamerGov! I left couple comments related to the padding implementation.

NarineK reviewed (commented):
I'm providing review comments for circuits, related test cases and related portion of tutorial. I'll look deeper into reducer and the new implementation of padding.

1. I think that it might be good to add more description on how those weights are computed, 
what are target1 and target2. With `output channel` I assume you meant the output of the layer target2 ?
2. When we call `show(W_3a_3b_hm)` can we explain color coding in the 5 x 5 matrix ?
3. As I was running the following cell I received an error message saying that: torchvision.transforms.transforms.RandomRotation is not a Module subclass.  I assume this is because I'm using an older version of torch vision. We might want to warn about that in the notebook. 
``` visualize_activations(model, model.mixed3a, 147) visualize_activations(model, 
      model.mixed3b, 379) ```
 I upgraded to torchvision 0.8 and it worked.
4. ... reduction alorgrithms can .. ->  reduction algorithms can ...
5. nit: `dectector` -> `detector` ? 
6. Could we perhaps de-abbreviate NMF and add a bit more description in the section:
    `Visualizing expanded weights`

ProGamerGov commented on captum/optim/_utils/circuits.py:
```diff
@@ -0,0 +1,41 @@
+    represented in a single weight tensor.
+
+    Schubert, et al., "Visualizing Weights", Distill, 2020.
+    See: https://distill.pub/2020/circuits/visualizing-weights/
```
Yeah, try this: https://drafts.distill.pub/distillpub/post--circuits-visualizing-weights/ with the password: weights

ProGamerGov replied:
The link is what I expect will be the link once it's no longer a draft.

ProGamerGov replied:
There's also: https://drafts.distill.pub/distillpub/post--circuits--frequency-edges/, which has the password: `hilo`.

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,106 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+    return activ_out
+
+
+class ActivationCatcher(object):
```
I forgot to mention: We usually do not extend from object. It applies by default in the latest versions of python.

ProGamerGov replied:
Should I extend from nn.module, or something else?

ProGamerGov replied:
I called it activation catcher because I was thinking about how the hooks 'catch' the input as it passes through them. But I can make it ActivationFetcher if that is a better name!

NarineK replied:
It doesn't quite look like a nn.Module. If we think that the model can be passed only once and reused, we can pass `model` as a constructor argument as well.

ProGamerGov replied:
Okay, I have it no longer extending from anything.

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,106 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+def collect_activations(
+    model,
+    targets: Union[nn.Module, List[nn.Module]],
+    model_input: torch.Tensor = torch.zeros(1, 3, 224, 224),
```
I'll make it a Union of Tensor and tuple of Tensors. I was also thinking the centering code might also have to a specified input, as not all models will have an NCHW output. But I'm not sure if that's something we need to do currently either?

ProGamerGov replied:
I've added the new type hint!

NarineK commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,106 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        assert sum(padding[6:]) == 0
+    if x.size(1) == 2:
+        assert sum(padding[4:6]) == 0
+
```
Long time ago I implemented circular padding in PT, I was wondering if we can adopt similar solution here so that we don't have to have if-elif for x.size(0).

https://github.com/NarineK/pytorch/blob/0a877cae6bf8e665b104d1c5a36c2a1c60f14a44/torch/nn/functional.py#L3020
I can try to adopt that solution for reflective padding.

NarineK replied:
There is in fact a more elegant solution with slice but it is not supported w/torch script
https://github.com/pytorch/pytorch/commit/6832149459c698ca2460670c6eb46e5948262e49#diff-c358c214f663ba0c8b9c6846fbe0042fa29494cf02fe4714a17dcd0d268b035bR3011

ProGamerGov replied:
@NarineK That be really awesome to not have the limitation! Are you able to implement it yourself?

ProGamerGov replied:
@NarineK Actually, I just got a reply back from Chris:

> Reflection would introduce artifacts (encourage the image to loop around at boundaries). You could do it, but I'd recommend against.
> Constant padding wouldn't be the end of the world.

So, it looks like we need symmetric to avoid artifacts.

NarineK replied:
Thank you for sharing the feedback, @ProGamerGov!

I'd actually recommend to convert the tensor into numpy, perform the padding in numpy and convert it back into tensor. You can do it by updating the data of the tensor and keeping other properties of the tensor such as `requires_grad`.
`tens.data` will give you the data and you can call numpy padding on it. 
It would make sense to make padding techniques in PT more sophisticated. It looks like currently the padding doesn't work when dim_size <= padding. We need to fix that in PyTorch.

NarineK replied:
`tens.data.cpu().numpy()` in case the tensor is on a GPU device.

ProGamerGov replied:
@NarineK Thanks for the idea! I was able to get it working with `torch.autograd.Function`, so SharedImage now uses NumPy's symmetric padding with preserving the gradient!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,106 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            return activations
+        except (Exception, BaseException) as e:
+            self.layers.remove_hooks()
+            raise e
```
Okay, I've made the change!

ProGamerGov commented on captum/optim/_utils/models.py:
```diff
@@ -151,3 +155,106 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+    return x
+
+
+def collect_activations(
```
Okay, I've moved it to `output_hook.py`.

ProGamerGov commented:
@NarineK I've added my first batch of changes based on your comments! I also re-added Ludwig's old center crop function as I needed it for the weight vis stuff to work correctly. It works differently than the CenterCrop transform, and I put in `transform.py`.

I'll work on changing the descriptions and text related stuff tomorrow.

I'm also not entirely sure what library or code I should use for telling if torchvision needs to be updated. In many of the tests, I check if the torch version matches a string to determine if the tests should be run or not.

ProGamerGov commented on tests/optim/utils/test_model_utils.py:
```diff
@@ -294,5 +334,40 @@ def test_get_layers_torchvision_alexnet(self) -> None:
+        try:
+            catch_activ = model_utils.ActivationCatcher(targets=[model.mixed4d])
+            activ_out = catch_activ(model, torch.zeros(1, 3, 224, 224))
+            self.assertIsInstance(activ_out, dict)
```
Okay, I've added an assert for the shapes!

ProGamerGov commented on captum/optim/_utils/circuits.py:
```diff
@@ -0,0 +1,41 @@
+    target2: nn.Module,
+    model_input: torch.Tensor = torch.zeros(1, 3, 224, 224),
+) -> torch.Tensor:
+    """
```
Okay, I've added a description of the arguments and return values!

ProGamerGov commented on captum/optim/_utils/circuits.py:
```diff
@@ -0,0 +1,41 @@
+def get_expanded_weights(
+    model,
+    target1: nn.Module,
+    target2: nn.Module,
```
Okay, I've clarified that target1 needs to come before target2 in the function's description!

ProGamerGov commented on tests/optim/utils/test_model_utils.py:
```diff
@@ -294,5 +334,40 @@ def test_get_layers_torchvision_alexnet(self) -> None:
+            activ_out = model_utils.collect_activations(
+                model, [model.mixed4d], torch.zeros(1, 3, 224, 224)
+            )
+            self.assertIsInstance(activ_out, dict)
```
I have an added an assert for the shapes!

ProGamerGov commented:
@NarineK I've updated the tutorial and function descriptions based on your comments, but I haven't added a note about the torchvision version yet. I also moved `tensor_heatmap` from the weight visualization notebook to `circuits.py` and wrote tests for it.

NarineK commented:
> @NarineK I've updated the tutorial and function descriptions based on your comments, but I haven't added a note about the torchvision version yet. I also moved `tensor_heatmap` from the weight visualization notebook to `circuits.py` and wrote tests for it, because I think that it'll be more useful that way.

Sounds good! Thank you, @ProGamerGov!
I was looking into structure of _utils package and thought that we can move image related things into a separate package.

I was looking into the codebase and it looks like there is another images.py in the _utils which contains only one function. I was wondering if it might make sense to create `image` folder under `_utils/image` and have `_utils/image/dataset.py`, `_utils/image/common.py` and   `_utils/image/reducer.py` (It looks like reducer contains `nchannels_to_rgb` code that is image specific we can move it into image and generalize later.)
The content of `_utils/images.py` can be moved under: `_utils/image/common.py`

`tensor_heatmap` function is also image specific, I'd move it to  `_utils/image/common.py`.

We might also consider moving `circuits.py` under `_utils/image/` because it has image flavor too.

ProGamerGov commented:
@NarineK ChannelReducer should work with any type of input I think, so we can leave it outside of `_utils/image`, but I'll move circuits.py into `_utils/image` until we are able to test it with non image models.

NarineK reviewed (commented):
Thank you for working on this PR, @ProGamerGov ! 
Posting comments about Channel Reducer ...

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+                permute_vals = [0] + list(range(x.ndim))[2:] + [1]
+                x = x.transpose(*permute_vals)
+
+        x_out = ChannelReducer._apply_flat(self._reducer.fit_transform, x)
```
@NarineK We don't need to convert it to a NumPy array as it seems to automatically be converted.

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def nchannels_to_rgb(x: torch.Tensor, warp: bool = True) -> torch.Tensor:
+    """
```
@NarineK Yeah, but it'll be a lot easier to use as a function in some cases. So, maybe I can create a wrapper class for `transform.py`? Should I leave the function in `utils/common.py` or should I also move it to `transform.py`?

NarineK replied:
I see  - If we have it as a transform how cumbersome will it be to use ? If we need a wrapper, we can keep the wrapper in the `_utils/image/common.py` and the transformation piece in the transform.py (if it is possible to represent as a transformation) ?

ProGamerGov replied:
The Lucid equivalent of nchannels_to_rgb is built into their `show()` function but sometimes it's called separately, and it's often used for displaying weights. If we make it class, users will have to use two lines of code instead of one to replicate the behavior. I also think don't Lucid ever uses it as a transform for optimizing model inputs with.

NarineK replied:
Sounds good, we can have the wrapper one line function to perform that transformation in the common.py.

ProGamerGov replied:
I know PyTorch has some of it's layer classes as wrappers for the functional equivalents, so I think it makes more sense to have the Class call the function as opposed to the other way around.

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+        self, n_components: int = 3, reduction_alg: Any = "NMF", **kwargs
+    ) -> None:
+        if not callable(reduction_alg):
+            algorithm_map = {}
```
@NarineK Future tutorials are going to be using other sklearn.decomposition functions like PCA, so we can't just make it NMF only.

NarineK replied:
@ProGamerGov , That was a typo, I didn't mean 'NMF' but name.

ProGamerGov replied:
@NarineK Ah, okay! I just skimmed the function and saw the NMF and assumed it might be NMF only.

NarineK replied:
Sorry about that - It was a typo

ProGamerGov replied:
Okay, I've made the change!

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+        self, x: Union[torch.Tensor, np.ndarray], reshape: bool = True
+    ) -> Union[torch.Tensor, np.ndarray]:
+        """
+        Perform dimensionality reduction on an input tensor or NumPy array.
```
There isn't a specific reason I guess, so I'll just make it PyTorch only.

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+
+    @classmethod
+    def _apply_flat(cls, func: Callable, x: torch.Tensor) -> torch.Tensor:
+        orig_shape = x.shape
```
I think we'll make it only a tensor for right now.

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+
+        if is_tensor:
+            if x.dim() == 3 and reshape:
+                x = x.permute(2, 1, 0)
```
The expected dimensionality is BCHW + optionally any number of additional dimensions, or CHW.

ProGamerGov replied:
I've added some clarification to the function as to what input formats are expected.

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+        return func(x.reshape([-1, x.shape[-1]])).reshape(list(orig_shape[:-1]) + [-1])
+
+    def fit_transform(
+        self, x: Union[torch.Tensor, np.ndarray], reshape: bool = True
```
I'll change the name to `swap_2nd_and_last_dims` for now, and we can see about adding the ability to specify where the channel dim is.

ProGamerGov commented on tests/optim/utils/test_reducer.py:
```diff
@@ -0,0 +1,160 @@
+        test_input = torch.randn(1, 32, 224, 224).abs()
+        c_reducer = reducer.ChannelReducer(n_components=3, max_iter=100)
+        test_output = c_reducer.fit_transform(test_input, reshape=True)
+        self.assertEquals(test_output.size(1), 3)
```
Okay, I've added asserts for the other dimensions!

ProGamerGov commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,192 @@
+    """
+
+    return torch.cat(
+        [torch.max(x, torch.full_like(x, 0)), torch.max(-x, torch.full_like(x, 0))],
```
@NarineK Yeah, I guess it is the same! Lucid probably did it that way because TensorFlow doesn't have functional layers or anything like that. Should we keep is as it is, or use ReLU?

NarineK replied:
@ProGamerGov, thank you for comparing with lucid. I think we can use ReLU since it is cleaner and probably easier to grasp.

ProGamerGov replied:
Okay, I've made the change!

ProGamerGov commented:
@NarineK So, I actually think it would be better to keep circuits.py outside of `_utils/image` for the moment. Other than that I've made all the suggested changes for `circuits.py` and `reducer.py`.

NarineK commented on captum/optim/_param/image/transform.py:
```diff
@@ -318,3 +369,45 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
+    """
+
+    @staticmethod
+    def forward(self, x: torch.Tensor, padding) -> torch.Tensor:
```
nit: type hint for padding ?

ProGamerGov replied:
Added!

NarineK commented on captum/optim/_param/image/transform.py:
```diff
@@ -318,3 +369,45 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return x
+
+    @staticmethod
+    def backward(self, grad_output: torch.Tensor):
```
is self here the context ? 
It might be more clear to store padding in the context and read from  there as in this example ?

https://pytorch.org/docs/stable/autograd.html?highlight=torch%20autograd%20function#torch.autograd.Function

ProGamerGov replied:
Yeah, I'll change `self` to `ctx`.

NarineK commented on captum/optim/_utils/circuits.py:
```diff
@@ -0,0 +1,63 @@
+    See: https://distill.pub/2020/circuits/visualizing-weights/
+
+    Args:
+        model:  PyTorch model instance.
```
nit: type hint: (nn.Module)
https://github.com/pytorch/captum/blob/master/captum/attr/_core/deep_lift.py#L112

ProGamerGov replied:
I've added the type hint!

NarineK commented on captum/optim/_utils/image/common.py:
```diff
@@ -0,0 +1,128 @@
+    )
+
+
+def tensor_heatmap(
```
nit: tensor sounds too generic. Perhaps, `weights_to_heatmap` ? I'm also open to other suggestions.

ProGamerGov replied:
`weights_to_heatmap_2d` may be more appropriate as the input has to have a dim of 2.

ProGamerGov replied:
I changed it to `weights_to_heatmap_2d`.

NarineK commented on captum/optim/_utils/image/common.py:
```diff
@@ -0,0 +1,128 @@
+
+def tensor_heatmap(
+    tensor: torch.Tensor,
+    colors: List[str] = ["0571b0", "92c5de", "f7f7f7", "f4a582", "ca0020"],
```
nit: In the public APIs we usually document all arguments as well.

ProGamerGov replied:
Okay, I've added args and the return info.

NarineK commented on captum/optim/_utils/reducer.py:
```diff
@@ -0,0 +1,112 @@
+        self, x: torch.Tensor, swap_2nd_and_last_dims: bool = True
+    ) -> torch.Tensor:
+        """
+        Perform dimensionality reduction on an input tensor or NumPy array.
```
nit: we might need to update the description about numpy array

ProGamerGov replied:
I've removed the part about NumPy arrays.

NarineK commented on tests/optim/param/test_images.py:
```diff
@@ -242,5 +247,316 @@ def test_laplacianimage_random_forward(self) -> None:
+                + [[abs(offset[3])] * 2]
+            )
+
+            x = SymmetricPadding.apply(x, offset_pad)
```
I think the test cases perform forward pass only. We'd need to call backward or autograd to test backward pass as well, don't we ?

ProGamerGov replied:
I've added a test for the backwards pass!

NarineK replied:
Thank you!

NarineK reviewed (commented):
Thank you for addressing all comments, @ProGamerGov! 
I left couple more nit comments. After we finish those, we can merge.

Before we merge to master later next year we'd need to add more description in the tutorials and more code documentation and testing. But we can do it in the later PRs.

NarineK commented on captum/optim/_utils/circuits.py:
```diff
@@ -23,7 +23,7 @@ def get_expanded_weights(
-        model:  PyTorch model instance.
+        model (nn.Module):  The reference to PyTorch model instance.
         target1 (nn.module):  The starting target layer. Must be below the layer
             specified for target2.
```
nit: I think following `nn.module` could be also updated to `nn.Module`

ProGamerGov replied:
Done!

NarineK commented on tests/optim/param/test_transforms.py:
```diff
@@ -391,6 +391,32 @@ def test_symmetric_padding(self) -> None:
+        x_out = sym_pad(x_pt, offset_pad)
+        (x_out.sum() * 1).backward()
+
+        assertTensorAlmostEqual(self, x, t_grad_input[0])
```
nit: could, please, we also assert `grad_output ` ?

ProGamerGov replied:
I've added an assert for grad_output!

NarineK replied:
Awesome! Thank you!

ProGamerGov commented:
@NarineK Alright, I've made the final fixes, so we can merge now! 

And I'll create a separate PR in the future for adding documentation our first alpha version of optim-wip!

NarineK commented:
> @NarineK Alright, I've made the final fixes, so we can merge now!!
> 
> And I'll create a separate PR in the future for adding documentation for our first alpha version of optim-wip!

Awesome! Thank you very much, @ProGamerGov ! Merging this PR!
